src/utils/dates.js

```diff
@@ -10,3 +12,101 @@ const ONE_DAY_IN_MILLISECONDS = 86400000;
 export function getFutureDate(offset) {
 	return new Date(Date.now() + offset * ONE_DAY_IN_MILLISECONDS);
 }
+
+export const handleNextPurchaseDate = (
+	name,
+	currentDate,
+	dateCreated,
+	purchasesCount,
+	dateNextPurchased,
+	dateLastPurchased,
+) => {
+	console.log('-------------------------------------------');
+	console.log(`[${name}]`);
+
+	try {
+		// get purchase intervals and get new estimation for next purchase date
+		const purchaseIntervals = calculatePurchaseIntervals(
+			currentDate,
+			dateCreated,
+			dateNextPurchased,
+			dateLastPurchased,
+		);
+		const estimatedNextPurchaseDate = getNextPurchaseEstimate(
+			purchaseIntervals,
+			purchasesCount,
+		);
+
+		console.log(
+			`New next purchase date: ${estimatedNextPurchaseDate.toLocaleString()}`,
+		);
+		console.log('-------------------------------------------');
+
+		return estimatedNextPurchaseDate;
+	} catch (error) {
+		throw new Error(`Failed getting next purchase date: ${error}`);
+	}
+};
+
+function getDaysBetweenDates(earlierDate, laterDate) {
+	return Math.floor(
+		(laterDate.getTime() - earlierDate.getTime()) / ONE_DAY_IN_MILLISECONDS,
+	);
+}
+
+function calculatePurchaseIntervals(
+	currentDate,
+	dateCreated,
+	dateNextPurchased,
+	dateLastPurchased,
+) {
+	const lastPurchaseDate = dateLastPurchased?.toDate(); // valid to be null
+	console.log(
+		`Previously purchased on: ${lastPurchaseDate === undefined ? undefined : lastPurchaseDate.toLocaleString()}`,
+	);
+	console.log(
+		`Previous next purchase date: ${dateNextPurchased.toDate().toLocaleString()}`,
+	);
+
+	// valid to be undefined
+	const lastEstimatedInterval = getDaysBetweenDates(
+		lastPurchaseDate === undefined ? dateCreated.toDate() : lastPurchaseDate,
+		dateNextPurchased.toDate(),
+	);
```

Comment:
Having a ternary within the function call is a bit difficult to read. I'd suggest breaking it out into another const like
```suggestion
        const lastEstimatedIntervalStartDate = lastPurchaseDate === undefined ? dateCreated.toDate() : lastPurchaseDate;
	const lastEstimatedInterval = getDaysBetweenDates(
		lastEstimatedIntervalStartDate,
		dateNextPurchased.toDate(),
	);
```

Reply:
Graet suggestion! decided to use nullish coalescing oparetor, because its so cool!

<code>const lastEstimatedIntervalStartDate = lastPurchaseDate ?? dateCreated.toDate();</code>